File: baekjoon/src/1012.py
# https://www.acmicpc.net/problem/1012


"""
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0]
[0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

방향벡터를 확인할 때, out or range 에러 발생을 방지하기 위해 field에 상하좌우에 한 줄의 패딩을 준다.
"""


def dfs(field, visited, coord):
    """
    1 1  => (1,1) (1,2)
      1           (2,2)

    (1,1)은 visited에 없으므로 검사가 필요하다. stack에 넣는다.
    (1,1)을 visited에 추가하고 방향벡터로 주변을 확인한다.
    1을 발견하면 visited에 없는지 확인하고 없다면 stack에 넣는다.

    (1,2)은 visited에 없으므로 이를 stack에 추가하여 위의 가정을 반복한다.
    (2,2)의 주변을 확인할 때 visited에 있거나 혹은 0이므로 넘어간다.
    stack이 비었으므로 반복을 종료한다.
    """
    dx, dy = [0, 0, -1, 1], [1, -1, 0, 0]
    stack = [coord]
    while stack:
        c = stack.pop()
        if c not in visited:
            visited.add(c)
            row, col = c
            for i in range(4):
                d_row, d_col = row + dx[i], col + dy[i]
                if field[d_row][d_col] == 0 or (d_row, d_col) in visited:
                    continue
                stack.append((d_row, d_col))


def solution():
    """
    M행 N열
    """
    M, N, K = map(int, input().split())
    field = [[0 for _ in range(M+2)] for _ in range(N+2)]
    for _ in range(K):
        col, row = map(int, input().split())
        field[row+1][col+1] = 1
    visited = set()
    count = 0
    for row in range(1, N + 1):
        for col in range(1, M + 1):
            cell = field[row][col]
            if cell == 0 or (row, col) in visited:
                continue
            dfs(field, visited, (row, col))
            count += 1
    return count


T = int(input())
for _ in range(T):
    print(solution())

File: baekjoon/src/1012.js
// https://www.acmicpc.net/problem/1012

const readline = require("readline");
const reader = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const lines = [];
reader
  .on("line", (line) => {
    lines.push(line.trim());
  })
  .on("close", () => {
    const T = Number(lines.shift());

    function getAroundCoords([x, y]) {
      return [
        [x - 1, y],
        [x, y - 1],
        [x + 1, y],
        [x, y + 1],
      ];
    }

    /**
     * 모든 곳을 확인해야한다. visited는 하나로 관리한다.
     *
     * 모든 구간을 순회하면서 더 이상 나아갈 수 없을 때까지 하고 주변에 갈 수 있는 곳이 아무 곳도 없으면 count를 1 증가한다.
     *
     * 1을 만나는 곳을 확인해서 visited에 없으면 그 구간에 한해서 dfs를 돌린다.
     */
    function solution() {
      const [M, N, K] = lines.shift().split(" ").map(Number);
      const items = [];
      for (let i = 0; i < K; ++i) items.push(lines.shift().split(" ").map(Number));

      const map = Array.from(Array(N), () => new Array(M).fill(0));
      for (const [col, row] of items) map[row][col] = 1;

      let count = 0;
      const visited = new Set();

      /**
       * 하나의 [row, col]을 받아서 주변을 확인하고 연결된 모든 좌표를 visited에 추가한다.
       */
      function dfs(coords, visited) {
        const stack = [coords];
        while (stack.length > 0) {
          const [r, c] = stack.pop();
          for (const [row, col] of getAroundCoords([r, c])) {
            const json = JSON.stringify([row, col]);
            if (!visited.has(json)) {
              visited.add(json);
              if (map[row] && map[row][col]) stack.push([row, col]);
            }
          }
        }
      }

      for (let row = 0; row < N; ++row) {
        for (let col = 0; col < M; ++col) {
          const flag = map[row][col];
          const json = JSON.stringify([row, col]);
          if (flag && !visited.has(json)) {
            visited.add(json);
            dfs([row, col], visited);
            count += 1;
          }
        }
      }

      console.log(count);
    }

    Array.from({ length: T }).forEach((_) => solution());
  });
